### PR TITLE
Require C++11.

### PR DESCRIPTION
**Pull request by drwells**
I spent some time this afternoon removing `DEAL_II_HAVE_CXX11` and `DEAL_II_WITH_CXX11` and simply issuing a `FATAL_ERROR` if the current checks are not met.

### CODE REVIEW

File: cmake/setup_finalize.cmake [hunk 1]
```diff
@@ -68,6 +68,10 @@ FOREACH(_feature ${DEAL_II_FEATURES})
   ENDIF()
 ENDFOREACH()
 
+# we require C++11:
+FILTER_SYSTEM_LIBRARIES(CXX11)
+REGISTER_FEATURE(CXX11)
```
Comment (tamiko) [inline cmake/setup_finalize.cmake]: The "feature" can be removed entirely. There are no libraries or headers or whatever attached to it.

**Comment by bangerth** [comment]
You must have started this right after @tjhei created the branch. That was apparently an issue that was of burning importance :-)

**Comment by tamiko** [comment]
@bangerth It is!
Same holds true for removing all the deprecated stuff superseded by `LinearOperator`.

My bet is that @drwells wants to set the record of highest negative  line count after a release branch :-)

**Comment by bangerth** [comment]
More power to him!

**Comment by bangerth** [comment]
I'm going to beat him by resolving #4065. But I need till tomorrow...

**Comment by bangerth** [comment]
(Patch mostly written already, though.)

**Comment by drwells** [comment]
@tamiko for a long time I had a net negative number of lines added to the library, but then I was foolish enough to write a bunch of tests and lost that honor :)

File: cmake/setup_finalize.cmake (hunk 1 above)
Comment (drwells) [inline cmake/setup_finalize.cmake]: Done.

**Comment by tamiko** [comment]
@drwells This pull request has a level of satisfaction that's hard to describe :-)

**Comment by drwells** [comment]
Writing it was very satisfying. There are a lot of nice things we will be able to do with `std::unique_ptr`, moves, lambdas (no more `bind`!), etc.

I don't really understand why the tester is failing to compile: there seems to be a mismatch between the language levels enabled by clang and the system headers. I would like to sort that out before this is merged (in addition to rebasing).

**Comment by drwells** [comment]
@davydden I put a list of features under the FAQ entry: what we require corresponds to the GCC 4.6 feature set (so, e.g., no `using` or `thread_local`).

I suppose that a related question is whether or not we want to require full C++11 compatibility or just compatibility for a subset. In this patch I have assumed that we will keep the subset requirement.

**Comment by davydden** [comment]
@drwells i am asking because of this issue https://github.com/dealii/dealii/issues/3209#issuecomment-276796858 where `FE_Enriched` used delegating constructors (`>=gcc4.7`). Current workaround is to disable the class unless `cxx14` is satisfied. If we agree to also require `delegating constructors` as a C++11 feature, then the `cxx14` check can be dropped and the class would already work on gcc4.8, for example.

**Comment by drwells** [comment]
Hm. We should settle this before we merge this PR. Lets do a poll:

@dealii/dealii :+1: this post if we should require everything in C++11: this would require Clang 3.4 or newer and GCC 4.8 or newer. :-1: this post if we should keep the current system.

If I recall correctly there are still a few contributors using GCC 4.6 for day to day work: for this reason I support keeping the current requirement. For the record I use RHEL 7 at work which does have GCC 4.8.

**Comment by drwells** [comment]
I believe that I fixed the problem that the tester ran into (the version flag was not propagated correctly). Things seem to work with GCC 6.3, clang 3.9, and GCC 4.7.

**Comment by drwells** [comment]
@dealii/dealii It looks like the 'require all of C++11' crowd is in the lead. Would anyone else like to vote on this?

**Comment by tjhei** [comment]
> If I recall correctly there are still a few contributors using GCC 4.6 for day to day work: for this reason I support keeping the current requirement.

One of my desktops is still running Ubuntu 12.04 with gcc 4.6. I don't mind too much and the machine will be upgraded soon, but I wonder how many people this change would effect.

If this is only about fe_enriched, we can either remove the usage of the delegating constructor or require cxx14 for fe_enriched.

**Comment by davydden** [comment]
> If this is only about fe_enriched, we can either remove the usage of the delegating constructor or require cxx14 for fe_enriched.

We already do the latter. Let's not make the code worse by removing C+11 features in 2017

**Comment by jppelteret** [comment]
I must admit that what I'd find frustrating is that we'd say that we support `C++11`, but we selectively only support a subset of it. Along the same lines as @davydden (I suppose), I feel that its a burden to maintain full support for [4](http://releases.llvm.org/) or [5](https://gcc.gnu.org/releases.html) year old compilers that "claim" to be standard conforming but are not. We're incrementing to a new major revision  - why not use the opportunity to make a clean break of it?

**Comment by drwells** [comment]
@jppelteret You convinced me to look at this more closely :)

It looks like Ubuntu 14.04 defaults to GCC 4.8 (as does RHEL 7) and Ubuntu 12.04 server edition will soon be EOLed (next month).

RHEL 6 defaults to GCC 4.4 which does not support our current C++11 subset and I don't think we can even compile with GCC 4.1 on the 8.5 branch right now (i.e., RHEL 5), so we should not be constrained by those.

AFAIK virtually every available version of clang supports all of C++11.

In conclusion, unless I am missing a distro, there is no strong case for supporting versions of GCC prior to 4.8 so I am changing my vote. IIRC GCC 4.8 and 4.9 don't have everything in `type_traits` but that is a minor issue we can handle with boost.

**Comment by tamiko** [comment]
With the mandatory C++11 requirement (what would that be anyway). I would like to ask everyone to calm down a bit and not jump the gun. We have for the last years maintained quite a bit of backwards compatibility. We should not throw that away lightly.

@jppelteret In fact, we "support" full C++11/14/17.
The question is more how much of the features we want to require (and ultimately what the lowest compiler version we support should be). And backwards compatibility to 4, or 5 years old compilers is not much at all.

For a long time we settled on gcc-4.6 (because wildly deployed on computing facilities and we had a dedicated tester for it). I agree that it is time to update our minimal required compiler versions. A sensible candidate could be gcc-4.8 (initial 2013, last minor revision 2015).

**Comment by davydden** [comment]
Another angle: we are speaking about 9.0 release, which roughly Spring 2018 (I.e. next year), that should also be taken into account

**Comment by bangerth** [comment]
I see a difference between requiring and using. When we say that we support C++11, that means that users can use C++11 features in their codes, but obviously only those that their compiler supports. It doesn't mean that we have to actively *use* all of these features in the library or the tutorials. I think the set of things that older C++11 compilers didn't fully support is pretty small (type traits, delegating constructors) that it's not overly burdensome to just not use them.

That said, I think it's ok to just use all of the things gcc 4.8 supported. The next release is not going to happen until late this year. If you still have too old a system by then, you may simply have to live with deal.II 8.5 until you move on to a newer system.

**Comment by jppelteret** [comment]
Although in context I'd agree that I should have said "require" instead of "support", in retrospect it does still seem appropriate. Doing my usual things and invoking a bit of philosophy, `deal.II` doesn't exist only as a library, but also an ecosystem of developers, contributors and users (who we hope might become contributors). For us to unequivocally allow contributions that use all of the `C++11` features (i.e. those that a modern compiler would not complain about), we lower the bar for contributing just a little and make the library just *that* much more accessible to our users to add to. Just a thought.

Following up of @bangerth's comment, and noting that talking about compilers and language specification is definitely not my forte, according to [these](https://gcc.gnu.org/projects/cxx-status.html) [pages](https://gcc.gnu.org/gcc-4.8/cxx0x_status.html),

> GCC 4.8.1 was the first feature-complete implementation of the 2011 C++ standard, previously known as C++0x.

so if we're going to be talking about having circa `GCC 4.8` as the minimum supported version, then I'd be in favour of this specific revision.

**Comment by drwells** [comment]
It sounds like the majority agree that we should require GCC 4.8 or newer and Clang 3.3 or newer which is (up to some minutia we can get around with boost and ABI details like the small string optimization) all of C++11. @tamiko @bangerth based on your comments I assume you are both in favor of this. @tjhei @kronbichler what, precisely, are your objections? I would like to hear the cons, if there are any, of this requirement before moving forward.

**Comment by kronbichler** [comment]
My objection is/was similar to what @bangerth and @tamiko raised: I am skeptical in breaking support for not-so-old systems without a specific need. I have several systems with RHEL 7 on them and gcc 4.8 installed. Even though I have compiled plenty of compilers myself, I know many users how would not do that, so we should support a system that is a few years old. A six year old compiler is definitely not that old.

The reason for downvoting was that gcc 4.8 does not contain *everything* of C++11, so I thought you might want even more than gcc 4.8, not the switch to enabling C++11 with the feature set that we have been using for some years now. But I am completely fine with requiring gcc 4.8 and similar.

**Comment by bangerth** [comment]
> For us to unequivocally allow contributions that use all of the |C++11|
> features (i.e. those that a modern compiler would not complain about),
> we lower the bar for contributing just a little and make the library
> just /that/ much more accessible to our users to add to.

Sure, but it's a trade-off because if we run off users who can't install 
the library on their systems, that's not helping either.

That said, I think we're over-debating the issue. The corner cases are 
few and far between. Let's just use whatever is in C++11 and if 
something breaks, we work around it. That's how we've always done it.

**Comment by drwells** [comment]
We may be overdebating but I wanted to make sure all sides were heard and I was not missing anything :). I'll modify the patch to check for a GCC 4.8 level of C++11 compliance. I'll post again when I am done.

**Comment by tamiko** [comment]
@drwells Please do not add checks that we currently do not need - the c++ checks are already horribly slow.

If you want gcc-4.8 compliance simply adjust the minimal compiler version required.

Actually, it might make sense to remove all of the checks and only keep regression checks by refactoring to `check_compiler_bugs.cmake`.

**Comment by drwells** [comment]
@tamiko My current plan is to use one test that uses multiple C++11 features that were not in GCC 4.7 as a general sanity check and then check known compiler versions.

We can throw out most of `check_compiler_bugs.cmake` since we don't support those compilers anymore :)

**Comment by drwells** [comment]
Clarification: I won't remove the old compiler bugs checks in this PR: that can wait until this is through.

**Comment by bangerth** [comment]
On 03/30/2017 02:14 PM, David Wells wrote:
> @tamiko <https://github.com/tamiko> My current plan is to use one test
> that uses multiple C++11 features that were not in GCC 4.7 as a general
> sanity check and then check known compiler versions.

I still wished we could come up with a way to run all of these checks in 
parallel somehow, given that they're all independent of each other. I 
know that cmake can't do this internally, but is it able to run multiple 
commands (each of which could itself be a mini cmake script) in parallel 
on separate processes? For example by creating one top-level target that 
depends on multiple sub-targets and then executing the makefile that 
corresponds to this with -j?

**Comment by drwells** [comment]
@bangerth I think we could do this with a shell script, but then we would lose Windows support. I combined a few of the C++ tests into one so we should still get a nice speed up.

This is ready for review again: the main changes are that we now require GCC 4.8 (on the todo list is figuring out other compiler minimum versions) and I removed most of the C++11 feature results. I kept the C++11 bug tests that apply to relatively recent systems: we may want to move those somewhere else in the future.

**Comment by bangerth** [comment]
This is ok with me, but I think @tamiko should look over it again.

**Comment by bangerth** [comment]
@tamiko ?

**Comment by tamiko** [comment]
\o/

@drwells I will do a couple of follow-up cleanups in `/lac` with respect to `linear_operator` and removing the obsolete matrix wrapping we had in `/lac` over the next days.

**Comment by bangerth** [comment]
Hurray!
